### PR TITLE
[Windows] Fix Button LineBreakMode property

### LINKED ISSUE #9277: [Windows] LineBreakMode doesn't work on Button
### Description

If you create a button which has text long enough that it should wrap, the Windows version won't break it into multiple lines.
image
Whereas on Android it is working just fine.
Screenshot_2022-08-09-00-03-20-137_com companyname mauiapp1

### Steps to Reproduce

Create a new Maui Project.
Replace the text of the button with something that should be broken into lines.
Observe the bug

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

Windows 11

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #10553: LineBreakMode cannot be used on a button on Windows
### Description

It looks like the `LineBreakMode` value on buttons is ignored on Windows. Here I have two identical buttons where the text doesn't fit, but the lower on has set `LineBreakMode` to `TailTruncation`:

image

But it does work on Android:

image

Also, note that you cannot see the beginning of the text on Windows.

### Steps to Reproduce

1. Create a new MAUI app.
2. Update MainPage.xaml:

``` XAML
<ScrollView>
    <VerticalStackLayout 
        Spacing="25" 
        Padding="30,0" 
        Margin="0,20,0,0"
        >

        <Button 
            Text="one two three four five six seven eight nine ten"
            MaximumWidthRequest="300"
            HorizontalOptions="Center" />

        <Button 
            Text="one two three four five six seven eight nine ten"
            MaximumWidthRequest="300"
            LineBreakMode="TailTruncation"
            HorizontalOptions="Center" />

    </VerticalStackLayout>
</ScrollView>
```

### Link to public reproduction project repository

https://github.com/pekspro/MauiIssues/tree/10553_LineBreakMode_cannot_be_used_on_a_button_on_Windows

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows, I was *not* able test on other platforms

### Affected platform versions

Windows 10.0.17763.0, Android 11

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Fix Windows Button LineBreakMode property.

fix-button-padding

image

### Issues Fixed

Fixes (but reverted) #9277 
Fixes #10553

### CODE REVIEW

**Comment by mattleibow** [comment]
@hartez I think you are working on big PR for all things line break and wrap... was that Android or everything? Will this conflict? Should we merge this and you are working on top?

**Comment by jsuarezruiz** [comment]
To help focus and move faster, we're going to have fewer open PRs and focus on the ones we'll be working on in each release. This PR is not included in the next release so I close it for now. I will reopen it and adapt to changes later.

**Review by mattleibow — COMMENTED** [review]
Is it possible to add tests for this? Maybe we can use the change in height to do something, or add some emoji with a color at the end and then when the ellipsis kicks in we can see that it is gone?

**Comment by jsuarezruiz** [comment]
> Is it possible to add tests for this? Maybe we can use the change in height to do something, or add some emoji with a color at the end and then when the ellipsis kicks in we can see that it is gone?

Sure thing.
image

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs [hunk 1]
```diff
@@ -5,11 +5,37 @@
...
 {
 	public partial class Button
 	{
+		WButton _button;
+
+		private protected override void OnHandlerChangedCore()
+		{
+			base.OnHandlerChangedCore();
+
+			if (Handler != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:21]:
```suggestion
			if (Handler is not null)
```

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs [hunk 2]
```diff
@@ -5,11 +5,37 @@
...
+			{
+				if (Handler is ButtonHandler buttonHandler && buttonHandler.PlatformView is WButton button)
+				{
+					_button = button;
+					_button.SizeChanged += OnButtonSizeChanged;
+				}
+			}
+			else
+			{
+				if (_button != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:31]:
```suggestion
				if (_button is not null)
```

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs [hunk 3]
```diff
@@ -33,5 +59,19 @@ public static void MapImageSource(IButtonHandler handler, Button button)
...
 			MapText((IButtonHandler)handler, button);
+
+		void OnButtonSizeChanged(object sender, SizeChangedEventArgs e)
+		{
+			if (Handler is ButtonHandler buttonHandler)
+			{
+				var platformView = buttonHandler.PlatformView;
+				var virtualView = buttonHandler.VirtualView as Button;
+
+				if (platformView == null || virtualView == null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:70]:
```suggestion
				if (platformView is null || virtualView is null)
```

File: src/Controls/src/Core/Platform/Windows/Extensions/ButtonExtensions.cs [hunk 4]
```diff
@@ -57,7 +57,27 @@ public static void UpdateLineBreakMode(this Microsoft.UI.Xaml.Controls.Button pl
...
+		internal static void UpdateContentSize(this Microsoft.UI.Xaml.Controls.Button platformButton, Button button)
+		{
+			if (platformButton.GetContent<TextBlock>() is TextBlock textBlock)
+			{
+				var buttonPadding = platformButton.Padding;
+				var horizontalSpace = buttonPadding.Left + buttonPadding.Right;
+
+				var image = platformButton.GetContent<UI.Xaml.Controls.Image>();
+
+				if (image != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/Windows/Extensions/ButtonExtensions.cs:73]:
```suggestion
				if (image is not null)
```

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs [hunk 5]
```diff
@@ -5,11 +5,37 @@
...
+				}
+			}
+			else
+			{
+				if (_button is not null)
+				{
+					_button.SizeChanged -= OnButtonSizeChanged;
+					_button = null;
+				}
+			}
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:36]: Lets reduce nested code as much as possible.
```suggestion
			else if (_button is not null)
			{
				_button.SizeChanged -= OnButtonSizeChanged;
				_button = null;
			}
```

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs [hunk 6]
```diff
@@ -33,5 +59,19 @@ public static void MapImageSource(IButtonHandler handler, Button button)
...
+		{
+			if (Handler is ButtonHandler buttonHandler)
+			{
+				var platformView = buttonHandler.PlatformView;
+				var virtualView = buttonHandler.VirtualView as Button;
+
+				if (platformView is null || virtualView is null)
+					return;
+
+				platformView.UpdateContentSize(virtualView);
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:73]: Idea, but take it or leave it if you want:
```suggestion
			if (Handler is ButtonHandler buttonHandler && buttonHandler.PlatformView is not null &&  buttonHandler.VirtualView is Button virtualView)
			{
				buttonHandler.PlatformView.UpdateContentSize(virtualView);
```

**Review by mandel-macaque — APPROVED** [review]
Minor changes, take them if you want or don't, as you please.

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs [hunk 7]
```diff
@@ -5,11 +5,37 @@
...
+using Microsoft.UI.Xaml;
+using WButton = Microsoft.UI.Xaml.Controls.Button;
 
 namespace Microsoft.Maui.Controls
 {
 	public partial class Button
 	{
+		WButton _button;
+
+		private protected override void OnHandlerChangedCore()
```
Comment (rmarinho) [inline src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:17]: Drop private

File: src/Controls/src/Core/Platform/Windows/Extensions/ButtonExtensions.cs [hunk 8]
```diff
@@ -57,7 +57,27 @@ public static void UpdateLineBreakMode(this Microsoft.UI.Xaml.Controls.Button pl
...
 			{
-				textBlock?.UpdateLineBreakMode(button.LineBreakMode);
+				textBlock.UpdateLineBreakMode(button.LineBreakMode);
+			}
+		}
+
+		internal static void UpdateContentSize(this Microsoft.UI.Xaml.Controls.Button platformButton, Button button)
+		{
+			if (platformButton.GetContent<TextBlock>() is TextBlock textBlock)
+			{
```
Comment (rmarinho) [inline src/Controls/src/Core/Platform/Windows/Extensions/ButtonExtensions.cs:67]: return if is not textblock

File: src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs (hunk 7 above)
Comment (jsuarezruiz) [reply src/Controls/src/Core/HandlerImpl/Button/Button.Windows.cs:17]: It is inherited, in order to do it I would need to change the virtual method in` Element.Impl.cs` and the rest of the overrides.

**Review by mattleibow — APPROVED** [review]

**Pull request merged**